api/sendfild/service/service.go

```diff
@@ -0,0 +1,713 @@
+	filter := bson.M{}
+
+	// Involving/from/to
+	if req.InvolvingFilter != "" {
+		filter["$or"] = bson.A{bson.M{"from": req.InvolvingFilter}, bson.M{"to": req.InvolvingFilter}}
+	} else {
+		if req.FromFilter != "" {
+			filter["from"] = req.FromFilter
+		}
+		if req.ToFilter != "" {
+			filter["to"] = req.ToFilter
+		}
+	}
+
+	// MessageState
+	if req.MessageStateFilter != pb.MessageState_MESSAGE_STATE_UNSPECIFIED {
+		filter["message_state"] = req.MessageStateFilter
+	}
+
+	// Waiting
+	if req.WaitingFilter != pb.WaitingFilter_WAITING_FILTER_UNSPECIFIED {
+		filter["waiting"] = req.WaitingFilter == pb.WaitingFilter_WAITING_FILTER_WAITING
+	}
+
+	ands := bson.A{}
+
+	// Amount eq/gte/lts/gt/lt
+	if req.AmountNanoFilEqFilter != 0 {
+		filter["amount_nano_fil"] = req.AmountNanoFilEqFilter
+	} else {
+		if req.AmountNanoFilGteqFilter != 0 {
+			ands = append(ands, bson.M{"amount_nano_fil": bson.M{"$gte": req.AmountNanoFilGteqFilter}})
+		} else if req.AmountNanoFilGtFilter != 0 {
+			ands = append(ands, bson.M{"amount_nano_fil": bson.M{"$gt": req.AmountNanoFilGtFilter}})
+		}
+
+		if req.AmountNanoFilLteqFilter != 0 {
+			ands = append(ands, bson.M{"amount_nano_fil": bson.M{"$lte": req.AmountNanoFilLteqFilter}})
+		} else if req.AmountNanoFilLtFilter != 0 {
+			ands = append(ands, bson.M{"amount_nano_fil": bson.M{"$lt": req.AmountNanoFilLtFilter}})
+		}
+	}
+
+	// Updated after/before
+	if req.UpdatedAfter != nil {
+		ands = append(ands, bson.M{"updated_at": bson.M{"$gt": req.UpdatedAfter.AsTime()}})
+	}
+	if req.UpdatedBefore != nil {
+		ands = append(ands, bson.M{"updated_at": bson.M{"$lt": req.UpdatedBefore.AsTime()}})
+	}
+
+	// Created after/before
+	if req.CreatedAfter != nil {
+		ands = append(ands, bson.M{"created_at": bson.M{"$gt": req.CreatedAfter.AsTime()}})
+	}
+	if req.CreatedBefore != nil {
+		ands = append(ands, bson.M{"created_at": bson.M{"$lt": req.CreatedBefore.AsTime()}})
+	}
+
+	// Apply paging info
+	comp := "$lt"
+	if req.MoreToken != 0 {
+		if req.Ascending {
+			comp = "$gt"
+		}
+		t := time.Unix(0, req.MoreToken)
+		ands = append(ands, bson.M{"created_at": bson.M{comp: &t}})
+	}
+
+	if len(ands) > 0 {
+		filter["$and"] = ands
+	}
+
+	cursor, err := s.col.Find(ctx, filter, findOpts)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "querying txns: %v", err)
+	}
+	defer cursor.Close(ctx)
+	var txns []txn
+	err = cursor.All(ctx, &txns)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "decoding txns query results: %v", err)
+	}
+
+	more := false
+	var startAt *time.Time
+	if len(txns) > 0 {
+		lastCreatedAt := &txns[len(txns)-1].CreatedAt
+		filter["created_at"] = bson.M{comp: *lastCreatedAt}
+		res := s.col.FindOne(ctx, filter)
+		if res.Err() != nil && res.Err() != mongo.ErrNoDocuments {
+			return nil, status.Errorf(codes.Internal, "checking for more data: %v", err)
+		}
+		if res.Err() != mongo.ErrNoDocuments {
+			more = true
+			startAt = lastCreatedAt
+		}
+	}
```

Comment:
I had to do this in other code I've been writing lately, would love to find a better way. But this runs an additional query to see if there is more data available so we can tell the client.

Reply:
I agree. Will change in upcoming PR.